hive-hooks/src/main/java/com/airbnb/reair/hive/hooks/MetastoreAuditLogHook.java

```diff
@@ -0,0 +1,382 @@
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.MetaStoreEventListener;
+import org.apache.hadoop.hive.metastore.api.MetaException;
+import org.apache.hadoop.hive.metastore.events.AddPartitionEvent;
+import org.apache.hadoop.hive.metastore.events.AlterPartitionEvent;
+import org.apache.hadoop.hive.metastore.events.AlterTableEvent;
+import org.apache.hadoop.hive.metastore.events.CreateDatabaseEvent;
+import org.apache.hadoop.hive.metastore.events.CreateTableEvent;
+import org.apache.hadoop.hive.metastore.events.DropDatabaseEvent;
+import org.apache.hadoop.hive.metastore.events.DropPartitionEvent;
+import org.apache.hadoop.hive.metastore.events.DropTableEvent;
+import org.apache.hadoop.hive.ql.hooks.ReadEntity;
+import org.apache.hadoop.hive.ql.hooks.WriteEntity;
+import org.apache.hadoop.hive.ql.hooks.WriteEntity.WriteType;
+import org.apache.hadoop.hive.ql.metadata.Partition;
+import org.apache.hadoop.hive.ql.metadata.Table;
+
+import com.airbnb.reair.db.DbCredentials;
+import com.airbnb.reair.utils.RetryableTask;
+import com.airbnb.reair.utils.RetryingTaskRunner;
+
+/**
+ * Audit logging for the metastore Thrift server. Comprises of a series of
+ * event listeners adding auditing to the various Trift event.
+ */
+public class MetastoreAuditLogHook extends MetaStoreEventListener {
+  public static Logger LOG = Logger.getLogger(MetastoreAuditLogHook.class);
+
+  // Number of attempts to make.
+  private static final int NUM_ATTEMPTS = 10;
+
+  // Will wait BASE_SLEEP * 2 ^ (attempt no.) between attempts.
+  private static final int BASE_SLEEP = 1;
+
+  public static String DB_USERNAME = "airbnb.reair.audit_log.db.username";
+  public static String DB_PASSWORD = "airbnb.reair.audit_log.db.password";
+
+  // Keys for values in hive-site.xml.
+  public static String JDBC_URL_KEY = "airbnb.reair.audit_log.jdbc_url";
+
+  protected DbCredentials dbCredentials;
+
+  /**
+   * Constructor which defines the relevant DB credentials.
+   *
+   * @param config The resource configuration
+   */
+  public MetastoreAuditLogHook(Configuration config) {
+    super(config);
+
+    dbCredentials = new ConfigurationDbCredentials(
+      getConf(),
+      DB_USERNAME,
+      DB_PASSWORD
+    );
+  }
+
+  /**
+   * Listener which fires when a table is created.
+   *
+   * The event only provides the table which was created, hence for auditing
+   * purposes the read/write differential is the non-presence and presence of
+   * the table respectively.
+   *
+   * @param event The create table event
+   */
+  @Override
+  public void onCreateTable(CreateTableEvent event) throws MetaException {
+    try {
+      Set<ReadEntity> readEntities = new HashSet<>();
+      Set<WriteEntity> writeEntities = new HashSet<>();
+
+      writeEntities.add(
+        new WriteEntity(
+          new Table(event.getTable()),
+          WriteType.INSERT
+        )
+      );
+
+      run(readEntities, writeEntities, HiveOperation.CREATETABLE);
```

Comment:
Since we're now creating an extended list of Hive operations ourselves, it would be cleaner to create new enums for the each API call type. Otherwise, the system will need to parse the string `THRIFT_API` in the command string.